File: C++/task3.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <cmath>
#include <stdexcept>
#include <iomanip>
// #include <Windows.h>

using namespace std;

double T;                   // Переменная для хранения начальной температуры кофе
double Ts;                  // Переменная для хранения температуры окружающей среды
double r;                   // Переменная для хранения коэффициента остывания
int time_limit;             // Переменная для хранения итогового времени
vector<double> apox_vector; // Вектор для хранения коэффициентов аппроксимирующей прямой

// Функция для вычисления коэффициента корреляции Пирсона
double korrel(const vector<double> &temperatures, double mean_y, int t)
{
    double sum_xy = 0, sum_x2 = 0, sum_y2 = 0;
    double mean_x = (time_limit - 1) / 2.0;
    for (int i = 0; i <= t; i++)
    {
        sum_xy += (i - mean_x) * (temperatures[i] - mean_y);
        sum_x2 += (i - mean_x) * (i - mean_x);
        sum_y2 += (temperatures[i] - mean_y) * (temperatures[i] - mean_y);
    }
    return sum_xy / sqrt(sum_x2 * sum_y2);
}

// Функция для вычисления коэффициентов аппроксимирующей прямой методом наименьших квадратов
pair<double, double> aprox(const vector<double> &x_vector, const vector<double> &y_vector)
{
    double sum_x = 0, sum_y = 0, sum_xy = 0, sum_x2 = 0;
    int n = x_vector.size();

    for (int i = 0; i < n; i++)
    {
        sum_x += x_vector[i];
        sum_y += y_vector[i];
        sum_xy += x_vector[i] * y_vector[i];
        sum_x2 += x_vector[i] * x_vector[i];
    }

    double a = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x);
    double b = (sum_y - a * sum_x) / n;

    return make_pair(a, b);
}

// Функция для моделирования остывания кофе и вычисления коэффициентов корреляции
vector<pair<double, double>> coffee(double T, double Ts, double r, int time_limit)
{
    vector<double> temperatures;
    vector<double> times;
    vector<pair<double, double>> temp_corr;

    for (int t = 0; t <= time_limit; t++)
    {
        double temperature = Ts + (T - Ts) * exp(-r * t);
        temperatures.push_back(temperature);
        times.push_back(t);
    }

    pair<double, double> aprox_coeffs = aprox(times, temperatures);
    double a = aprox_coeffs.first;
    double b = aprox_coeffs.second;

    apox_vector = {a, b};

    double mean_y = accumulate(temperatures.begin(), temperatures.end(), 0.0) / temperatures.size();
    for (int t = 0; t <= time_limit; t++)
    {
        double corr = korrel(temperatures, mean_y, t);
        temp_corr.push_back(make_pair(temperatures[t], corr));
    }

    return temp_corr;
}

int main()
{
    // SetConsoleCP(1251);
    // SetConsoleOutputCP(1251);

    cout << "Введите начальную температуру кофе: ";
    cin >> T;
    cout << "Введите температуру окружающей среды: ";
    cin >> Ts;
    cout << "Введите коэффициент остывания: ";
    cin >> r;
    cout << "Введите итоговое время: ";
    cin >> time_limit;

    // Вызов функции coffee для моделирования остывания кофе
    vector<pair<double, double>> results = coffee(T, Ts, r, time_limit);

    cout << "Коэффициенты аппроксимирующей прямой а - " << apox_vector[0] << ", b - " << apox_vector[1] << endl;
    cout << "+------+" << "------------+" << "---------------+" << endl;
    cout << "| Время|" << " Температура|" << " Коэф.кореляции|" << endl;
    cout << "+------+" << "------------+" << "---------------+" << endl;

    int t = 0;

    // Цикл для вывода результатов моделирования остывания кофе
    for (const auto &result : results)
    {

        cout << "|" << setw(6) << t << "|" << setw(12) << result.first << "|" << setw(15) << result.second << "|" << endl;
        t++;
        cout << "+------+" << "------------+" << "---------------+" << endl;
    }

    return 0;
}

File: rust/task3.rs
use std::f64::consts::E;
use std::io;
use std::iter::Iterator;

fn korrel(temperatures: &Vec<f64>, mean_y: f64, t: usize) -> f64 {
    let mut sum_xy = 0.0;
    let mut sum_x2 = 0.0;
    let mut sum_y2 = 0.0;
    let mean_x = (t as f64) / 2.0;
    for i in 0..=t {
        sum_xy += (i as f64 - mean_x) * (temperatures[i] - mean_y);
        sum_x2 += (i as f64 - mean_x) * (i as f64 - mean_x);
        sum_y2 += (temperatures[i] - mean_y) * (temperatures[i] - mean_y);
    }
    sum_xy / (sum_x2 * sum_y2).sqrt()
}

fn aprox(x_vector: &Vec<f64>, y_vector: &Vec<f64>) -> (f64, f64) {
    let n = x_vector.len();
    let mut sum_x = 0.0;
    let mut sum_y = 0.0;
    let mut sum_xy = 0.0;
    let mut sum_x2 = 0.0;
    for i in 0..n {
        sum_x += x_vector[i];
        sum_y += y_vector[i];
        sum_xy += x_vector[i] * y_vector[i];
        sum_x2 += x_vector[i] * x_vector[i];
    }
    let a = (n as f64 * sum_xy - sum_x * sum_y) / (n as f64 * sum_x2 - sum_x * sum_x);
    let b = (sum_y - a * sum_x) / (n as f64);
    (a, b)
}

fn coffee(T: f64, Ts: f64, r: f64, time_limit: usize) -> Vec<(f64, f64)> {
    let mut temperatures = Vec::new();
    let mut times = Vec::new();
    let mut temp_corr = Vec::new();
    for t in 0..=time_limit {
        let temperature = Ts + (T - Ts) * E.powf(-r * t as f64);
        temperatures.push(temperature);
        times.push(t as f64);
    }
    let (a, b) = aprox(&times, &temperatures);
    let mean_y = temperatures.iter().sum::<f64>() / temperatures.len() as f64;
    for t in 0..=time_limit {
        let corr = korrel(&temperatures, mean_y, t);
        temp_corr.push((temperatures[t], corr));
    }
    temp_corr
}

fn main() {
    let mut T: f64 = 0.0;
    let mut Ts: f64 = 0.0;
    let mut r: f64 = 0.0;
    let mut time_limit: usize = 0;

    println!("Введите начальную температуру кофе: ");
    T = read_input();
    println!("Введите температуру окружающей среды: ");
    Ts = read_input();
    println!("Введите коэффициент остывания: ");
    r = read_input();
    println!("Введите итоговое время: ");
    time_limit = read_input_usize();

    let results = coffee(T, Ts, r, time_limit);

    let (a, b) = aprox(&(0..=time_limit as isize).map(|x| x as f64).collect(), &results.iter().map(|x| x.0).collect());
    println!("Коэффициенты аппроксимирующей прямой a - {}, b - {}", a, b);
    println!("+------+------------+---------------+");
    println!("| Время| Температура| Коэф.кореляции|");
    println!("+------+------------+---------------+");
    for (t, result) in results.iter().enumerate() {
        println!("|{:6}|{:12.4}|{:15.4}|", t, result.0, result.1);
        println!("+------+------------+---------------+");
    }
}

fn read_input() -> f64 {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read input");
    input.trim().parse().expect("Invalid input")
}

fn read_input_usize() -> usize {
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read input");
    input.trim().parse().expect("Invalid input")
}
